File: Adam_FastQ_BaseQual_Baseline/src/main/scala/FastQ_PerBaseQual_Reducer.scala
package main.scala

import main.utils.AvgCount

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 04.02.16.
  */
object FastQ_PerBaseQual_Reducer {

  def createAverageCount(qualVal: Int): AvgCount = {
    new AvgCount(qualVal,1)
  }

  def addAndCount(a: AvgCount, qualVal: Int): AvgCount = {
    a.addQualVal(qualVal)
  }

  def combine(a: AvgCount, b: AvgCount): AvgCount = {
    new AvgCount(a.getSumOfQualityValues()+b.getSumOfQualityValues(), a.getNumberOfQualityValues()+b.getNumberOfQualityValues())
  }

}

File: MapReduce_FastQ_BaseQual_Optimized/src/main/java/baseQual/FastQ_PerBaseQual_Reducer.java
package baseQual;

import org.apache.hadoop.io.DoubleWritable;
import utils.IdentifierPositionKeyWritable;
import utils.QualityCountHelperWritable;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 21.04.2016
 */
public class FastQ_PerBaseQual_Reducer extends org.apache.hadoop.mapreduce.Reducer<IdentifierPositionKeyWritable, QualityCountHelperWritable, IdentifierPositionKeyWritable, DoubleWritable> {

    @Override
    protected void reduce(IdentifierPositionKeyWritable key, Iterable<QualityCountHelperWritable> values, Context context) throws IOException, InterruptedException {
        context.write(key, getMeanValue(values));
    }

    private DoubleWritable getMeanValue(Iterable<QualityCountHelperWritable> values) {
        int sumOfQualityValues = 0;
        int numberOfQualityValues = 0;
        for (QualityCountHelperWritable qualCounter : values) {
            sumOfQualityValues += qualCounter.getSumOfQualityValues();
            numberOfQualityValues += qualCounter.getNumberOfQualityValues();
        }
        return new DoubleWritable((double)sumOfQualityValues / (double)numberOfQualityValues);
    }

}